design/tricky-choices/dag-pb-forms-impl-and-use.md

```diff
@@ -0,0 +1,126 @@
+
+But, because it still does proper data model form path resolution, you can retrieve CIDs via `Qm.../Links/X/Hash` (where `X` is an index of the `Links` array) and it will resolve properly to linked block. So JavaScript gives us two ways to resolve links in dag-pb: `Qm.../NamedLink` and `Qm.../Links/X/Hash`. go-ipfs gives us `Qm.../NamedLink`.
+
+After js-ipfs v0.56.0, which contained the upgrade to the newer js-multiformats IPLD stack that uses much more plain object forms for representing the data model, **both pathing options now no longer appear to work**, although it still returns `{"links":[]}` if the path should be valid, suggesting that resolution is working but we're not seeing the nodes, but that CIDs are not being followed. There are some bugs here that we should chase down.
+
+## go-ipld-prime in go-ipfs
+
+As of writing, the go-ipld-prime in go-ipfs work as represented in https://github.com/ipfs/go-ipfs/pull/7976 changes all of this for go-ipfs.
+
+1. go-merkledag is no longer used in the primary `dag get` and `dag put` path as far as IPLD block handling and pathing are concerned.
+2. go-codec-dagpb is used to instantiate full `{Data,Links:[{Name,Tsize,Hash},...]}` forms, [as per the IPLD Schema in the dag-pb spec](https://ipld.io/specs/codecs/dag-pb/spec/#logical-format) (matching the protobuf schema).
+3. go-ipld-prime handles path resolution / traversals for `dag get`
+4. There is special-casing, via go-unixfsnode, such that loading a dag-pb node will attempt to reify it into a [UnixFS data model form which has a custom `LookupByString`](https://github.com/ipfs/go-unixfsnode/blob/main/pathpbnode.go) that will search its dag-pb node `Links` array for a named link and return that CID if it has one by that name.
+5. The custom `LookupByString` means that you can't resolve anything else in the data model form of the dag-pb node.
+6. There is an `/ipld/` override for resolution, such that when the path is prefixed with this, will _not_ perform the unixfsnode reification and you'll get the "pure" data model form and can resolve all of the properties of `{Data,Links:[{Name,Tsize,Hash},...]}` in a standard pathing form.
+
+Worth noting is that the current (pre go-ipld-prime) version of go-ipfs **does not do anything special with the `/ipld/` path prefix and dag-pb nodes**. It appears to behave the same as without, resolving named links and nothing else.
```

Comment:
This is close, but I don't think quite accurate. IIUC what's happening both pre and post go-ipld-prime with respect to /ipfs and /ipld is the following:

Pre go-ipld-prime:
- /ipfs -> Interpret the data as UnixFS, meaning QmFoo/NamedItemInAHAMTDirectory/bar` works
- /ipld -> Interpret the data as non-IPLD, but with ["legacy pathing"](https://github.com/ipld/specs/blob/a652a23ba974c7b524882602a12b6d978b13829a/block-layer/codecs/dag-pb.md#alternative-legacy-pathing) that traverses the named links and hides the size and data fields. In go-ipld-prime world this is analogous to using [PathPBNode](https://github.com/ipfs/go-unixfsnode/blob/d6733f6db69431075c9cf7b30e0ee1f6a082f4b0/pathpbnode.go) as a lens.

Post go-ipld-prime (branch today):
- /ipfs -> Interpret the data as UnixFS
- /ipld -> Interpret the data as plain as possible without any legacy pathing

Main takeaways from an investigative view (i.e. not judging the outcomes):
- /ipld did used to do something in go-ipfs, go-ipld-prime is changing the behavior
- Keeping /ipfs is not a reliable way for us to do named pathing through DagPB objects since we'll use the UnixFS children instead of the object links (e.g. no accessing of named HAMT shard links). Additionally, if we add more strictness to the UnixFS reifier then even basic non-UnixFS DagPB nodes might be untraversable this way.
- As it stands with go-ipld-prime we have killed off named path traversal of DagPB objects via `ipfs dag get`

Reply:
I don't quite understand the last two points you have here @aschmahmann.

From what I understand, and from experimenting, it looks like `QmFoo/NamedLink/foo` _and_ `QmFoo/NamedShardedLink/foo` works old and new, it's just that the output is now dag-json and has the TitleCased `Data` and `Links`. Aside from that, the difference is in `/ipld`, as you've pointed out, works for named links with the old, but not sharded named links. I can do `/ipld/QmFoo/NamedLink` but not `/ipld/QmFoo/NamedShardedLink` but I can do `/ipld/QmFoo/AA`, `AA` being one of the names that the HAMT is going to give for its sharded named links.

So with the go-ipld-prime changes, the only loss (aside from json changes) is the ability to use unsharded named links on sharded directories (`AA` etc.). But you still have the ability to access those by index, so `/ipld/QmFoo/Links/170` is going to get me the same as `/ipld/QmFoo/AA` would have got me before.

Which is all pretty obscure and doesn't seem like a loss of functionality for the vast majority of users, likely no users since this is all hidden (including behind a flag in go-ipfs!).

I'm going to update the doc to talk about sharded directories because that's a difference I didn't investigate yesterday.